packages/selected-network-controller/src/SelectedNetworkController.ts

```diff
@@ -153,60 +155,51 @@ export class SelectedNetworkController extends BaseController<
     this.#domainProxyMap = domainProxyMap;
     this.#registerMessageHandlers();
 
-    // this is fetching all the dapp permissions from the PermissionsController and looking for any domains that are not in domains state in this controller. Then we take any missing domains and add them to state here, setting it with the globally selected networkClientId (fetched from the NetworkController)
+    const selectedChainId = this.#getSelectedChainId();
+    // this is fetching all the dapp permissions from the PermissionsController and looking for any domains that are not in domains state in this controller. Then we take any missing domains and add them to state here, setting it with the globally selected chainId (fetched from the NetworkController)
     this.messagingSystem
       .call('PermissionController:getSubjectNames')
       .filter((domain) => this.state.domains[domain] === undefined)
-      .forEach((domain) =>
-        this.setNetworkClientIdForDomain(
-          domain,
-          this.messagingSystem.call('NetworkController:getState')
-            .selectedNetworkClientId,
-        ),
-      );
+      .forEach((domain) => this.setChainIdForDomain(domain, selectedChainId));
 
     this.messagingSystem.subscribe(
       'PermissionController:stateChange',
       (_, patches) => {
+        const selectedChainId = this.#getSelectedChainId();
         patches.forEach(({ op, path }) => {
           const isChangingSubject =
             path[0] === 'subjects' && path[1] !== undefined;
           if (isChangingSubject && typeof path[1] === 'string') {
             const domain = path[1];
             if (op === 'add' && this.state.domains[domain] === undefined) {
-              this.setNetworkClientIdForDomain(
-                domain,
-                this.messagingSystem.call('NetworkController:getState')
-                  .selectedNetworkClientId,
-              );
+              this.setChainIdForDomain(domain, selectedChainId);
             } else if (
               op === 'remove' &&
               this.state.domains[domain] !== undefined
             ) {
-              this.#unsetNetworkClientIdForDomain(domain);
+              this.#unsetChainIdForDomain(domain);
             }
           }
         });
       },
     );
 
+    // TODO also monitor the defaultRpcEndpoint for a chainId changing
```

Comment:
Is this something that needs to be done?  Feels like the only thing that needs to be reacted to is the whole network/chain being deleted.

Reply:
If migrating to chain ids is too much effort in the short term, this change achieves a similar effect by redirecting domains to the default endpoint when it changes: https://github.com/MetaMask/core/commit/10fbdf1203faa16d4eccb0693b2a2b9b8f2bbb74

Reply:
If moving to chain ids is too much effort in the short term, a similar effect can be achieved by redirecting domains to the default endpoint when it changes: https://github.com/MetaMask/core/commit/0613428bf098dba76fa2219e13326be05148ca1a

Reply:
Hey @bergeron yeah Jiexi and I strategized around this and decided to pursue both options. 
"redirecting domains to the default endpoint when it changes" per your and my conversation at first - in case the work in this PR takes a bit longer to merge.

Reply:
> in the extension

Do you have more precise idea for how this would work?

Here's the 2 reasons I implemented it via #4679 instead of through the extension.

(1) The `SelectedNetworkController` will need changes either way, since the current `NetworkController:stateChange` [subscriber](https://github.com/MetaMask/core/blob/fe3dfcb5f71803c103ab7e6d91f91bcac2ae9c3c/packages/selected-network-controller/src/SelectedNetworkController.ts#L203) doesn't do what we want.  Deleting a single endpoint currently gets it defaulted to the global network client instead of the network's default.  So we'll be editing that subscriber anyway.

(2) This case where you both delete the default endpoint, and replace it with a new default in 1 operation to `updateNetwork`:

https://github.com/user-attachments/assets/fcf8e16b-1424-418f-8bc7-5cb54f492383

The extension doesn't know the new endpoint's network client id until the controller autogenerates it.  So domain edits have to occur in response to the state update, and the `SelectedNetworkController`'s seemed most convenient as it has everything to handle both removes and edits in 1 place.

Reply:
> (1) The SelectedNetworkController will need changes either way, since the current NetworkController:stateChange [subscriber](https://github.com/MetaMask/core/blob/fe3dfcb5f71803c103ab7e6d91f91bcac2ae9c3c/packages/selected-network-controller/src/SelectedNetworkController.ts#L203) doesn't do what we want. Deleting a single endpoint currently gets it defaulted to the global network client instead of the network's default. So we'll be editing that subscriber anyway.

I suppose this does make it tricky. I was thinking we could subscribe to this event in the MetaMask-Controller and call `setNetworkClientDomain` using the SelectedNetworkController's public API (similar pattern to what we do in several ways [here](https://github.com/MetaMask/metamask-extension/blob/9405bc1dbccd4c51cff60b32a0386ff3ed55b4b2/app/scripts/metamask-controller.js#L2740)) But if we're competing with the subscription already in the controller we could have some funky race conditions...